### PR TITLE
gh-117549: Match declaration order for _Py_BackoffCounter initializer

### PR DESCRIPTION
**Pull request by mdboom**
This is to fix the build for Scipy, which includes this header from C++ code which seems to really care about such things.

To be clear, I didn't confirm that this fixes the Scipy build -- there are a bunch of other patches required to its build system to even compile with Python 3.13 and I'm not sure it's worth the effort there.

But this warning seems legitimate and the fix is harmless.  It's possible there are other errors in Scipy behind it, but this is the only struct initializer added in #117144.

* Issue: gh-117549

### CODE REVIEW

**Review by gvanrossum — APPROVED** [review]
LGTM. Let's wait for @tacaswell to give us a thumbs up here before we merge.

**Comment by tacaswell** [comment]
I can confirm that scipy compiles with cpython from this branch.

**Review by tacaswell — APPROVED** [review]

**Comment by gvanrossum** [comment]
Thanks for your help! Merging now...

**Comment by methane** [comment]
Cython includes this file and g++ 9 and 10 (available on Ubuntu 22.04 LTS) doesn't compile this.

https://github.com/msgpack/msgpack-python/issues/593#issuecomment-2067409274

```cpp
// foo.cpp
#define Py_BUILD_CORE 1
#include <Python.h>

#include "frameobject.h"
#include "internal/pycore_frame.h"

int main() {}
// g++-10 -std=c++17 -I include/python3.13 foo.cpp
```

I confirmed Using temporary variable fix it.

```c
    _Py_BackoffCounter c;
    c.backoff = backoff;
    c.value = value;
    return c;
```

But I am not sure we should support g++ 9/10 in Python 3.13.

**Comment by gvanrossum** [comment]
I am despairing. We make all these efforts to indicate things are internal and they are still included in half the packages in the world. :-(

**Comment by methane** [comment]
FYI, this caused compile error on MSVC too. Should we add `/std:c++20` in cflags on msvc for better interoperability with cython&c++?
https://github.com/msgpack/msgpack-python/issues/597

**Comment by gvanrossum** [comment]
Wait— if it failed on MSVC how could the original code have passed CI?

**Comment by methane** [comment]
Because we don't test C++.
This is problem only when using C++ with Cython.

**Comment by methane** [comment]
`/std:c++20` didn't help me. I need to stop using C++ or Cython.

```
  C:\hostedtoolcache\windows\Python\3.13.0-alpha.6\x64\include\internal\pycore_backoff.h(47): error C4576: a parenthesized type followed by an initializer list is a non-standard explicit type conversion syntax
  C:\hostedtoolcache\windows\Python\3.13.0-alpha.6\x64\include\internal\pycore_backoff.h(53): error C4576: a parenthesized type followed by an initializer list is a non-standard explicit type conversion syntax
```

**Comment by methane** [comment]
```
diff --git a/Include/internal/pycore_backoff.h b/Include/internal/pycore_backoff.h
index decf92bc419..6ac69b52168 100644
--- a/Include/internal/pycore_backoff.h
+++ b/Include/internal/pycore_backoff.h
@@ -44,13 +44,15 @@ make_backoff_counter(uint16_t value, uint16_t backoff)
 {
     assert(backoff <= 15);
     assert(value <= 0xFFF);
-    return (_Py_BackoffCounter){.backoff = backoff, .value = value};
+    _Py_BackoffCounter ret = {.backoff = backoff, .value = value};
+    return ret;
 }

 static inline _Py_BackoffCounter
 forge_backoff_counter(uint16_t counter)
 {
-    return (_Py_BackoffCounter){.as_counter = counter};
+    _Py_BackoffCounter ret = {.as_counter = counter};
+    return ret;
 }
```

**Comment by gvanrossum** [comment]
So let’s fix it. Do you have time for a PR?

**Comment by gvanrossum** [comment]
That should be fixed now.

**Comment by methane** [comment]
thank you.

**Comment by methane** [comment]
By the way, Cython uses this only for building custom traceback object.
We hided `f_lineno` in PyFrameObject so Cython includes `pycore_frame.h`.

Adding `_PyFrame_SetLineno()` in `Include/cpython/frameobject.h` makes it possible to Cython remove this include.

https://github.com/cython/cython/blob/45db483cbe128c29a1e9ffe9f5a4db12cf77cddd/Cython/Utility/Exceptions.c#L1007C1-L1007C52
https://github.com/cython/cython/blob/45db483cbe128c29a1e9ffe9f5a4db12cf77cddd/Cython/Utility/ModuleSetupCode.c#L839

**Comment by gvanrossum** [comment]
> Adding `_PyFrame_SetLineno()` in `Include/cpython/frameobject.h` makes it possible to Cython remove this include.

So IIUC you are proposing to add a new public API to set the `f_lineno` attribute in a frame object? It's too bad that today is the 3.13 beta 1 feature freeze, so this will have to go into 3.14. Can you open a new issue and PR so we won't forget?

**Pull request merged**
